### PR TITLE
The Great TypeScript Refactor

### PR DESCRIPTION
**Pull request by 0xdavinchee**
## Overview
- Test files converted to `.ts`
- Type files added
- Separate linting for .js and .ts files
- Types added accordingly
- Typechain added for contract types

### CODE REVIEW

File: .github/workflows/cd.feature.create-pr-artifact.yml [hunk 1]
```diff
@@ -63,7 +63,7 @@ jobs:
           prNumber=${{ github.event.number }}
           dirName="bin/${prNumber}"
           mkdir -p $dirName
-          for d in packages/ethereum-contracts packages/js-sdk packages/sdk-core packages/sdk-redux; do (
+          for d in packages/ethereum-contracts packages/sdk-core packages/sdk-redux; do (
```
Comment (0xdavinchee) [inline .github/workflows/cd.feature.create-pr-artifact.yml:66]: The .yml changes are almost entirely just removing js-sdk from it.

File: package.json [hunk 2]
```diff
@@ -13,10 +13,10 @@
         "git:submodule:cleanup": "git submodule deinit --all --force",
         "lint": "yarn workspaces run lint",
         "build": "yarn build-essentials #synonym of build-essentials",
-        "build-essentials": "set -e;for i in ethereum-contracts js-sdk sdk-core sdk-redux;do yarn workspace @superfluid-finance/$i build;done",
-        "build-for-contracts-dev": "set -e;for i in ethereum-contracts js-sdk;do yarn workspace @superfluid-finance/$i build;done",
+        "build-essentials": "set -e;for i in ethereum-contracts sdk-core sdk-redux;do yarn workspace @superfluid-finance/$i build;done",
```
Comment (0xdavinchee) [inline package.json]: js-sdk package removed from root package.json

File: packages/ethereum-contracts/package.json [hunk 3]
```diff
@@ -37,7 +37,7 @@
         "testenv:deploy": "yarn run-truffle exec scripts/deploy-test-environment.js : TEST",
         "pretest": "{ yarn testenv:start > /dev/null& } && sleep 5",
         "test": "run-s test:*:*",
-        "test:contracts:hardhat": "yarn run-hardhat test testsuites/all-contracts.js",
+        "test:contracts:hardhat": "yarn run-hardhat test testsuites/all-contracts.ts",
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/package.json:40]: ts time now

File: packages/ethereum-contracts/package.json [hunk 4]
```diff
@@ -58,7 +58,7 @@
     "dependencies": {
         "@decentral.ee/web3-helpers": "0.5.3",
         "@openzeppelin/contracts": "4.7.3",
-        "@superfluid-finance/js-sdk": "0.6.3",
+        "@superfluid-finance/js-sdk": "0.6.2",
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/package.json:61]: 0.6.3 is unreleased and not a package in the monorepo anymore

**Comment by hellwolf** [comment]
wot, great!

**Comment by hellwolf** [comment]
image

okay :/ fine

File: packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.ts [hunk 5]
```diff
@@ -1,52 +1,65 @@
-const TestEnvironment = require("../../TestEnvironment");
-const {expectCustomError} = require("../../utils/expectRevert");
-const {expectEvent} = require("@openzeppelin/test-helpers");
-const {ethers} = require("hardhat");
-const {expect} = require("chai");
-const {toBN, toWad, max} = require("../utils/helpers");
-const {
+import {SignerWithAddress} from "@nomiclabs/hardhat-ethers/signers";
```
Comment (hellwolf) [inline packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.ts]: while we are at it, let's split this file into multiple ones: testing without callbacks, testing with MFA, and testing with callbacks.

File: packages/ethereum-contracts/test/contracts/agreements/InstantDistributionAgreementV1.behaviour.ts [hunk 6]
```diff
@@ -1,21 +1,29 @@
...
+import {
+    IDABaseParams,
+    IDAIndexData,
+    IDASubscriptionData,
+} from "./Agreement.types";
+
 const {web3tx, toBN, wad4human} = require("@decentral.ee/web3-helpers");
+const {expectEvent} = require("@openzeppelin/test-helpers");
 
 function _updateIndexData({
```
Comment (hellwolf) [inline packages/ethereum-contracts/test/contracts/agreements/InstantDistributionAgreementV1.behaviour.ts:13]: so these object-ish parameters would not be typed?

**Comment by hellwolf** [comment]
anything left?

**Comment by 0xdavinchee** [comment]
> anything left?

no, it's good to go

**Review by hellwolf — APPROVED** [review]
lfg

**Pull request merged**